Userland/Libraries/LibThreading/BackgroundAction.h

```diff
@@ -34,51 +35,72 @@ class BackgroundActionBase {
 };
 
 template<typename Result>
-class BackgroundAction final : public Core::Object
-    , private BackgroundActionBase {
-    C_OBJECT(BackgroundAction);
+struct ResultWrapper : public Core::Object {
+    C_OBJECT(ResultWrapper);
 
-public:
-    void cancel()
+    ResultWrapper(Result& result)
+        : result(result)
     {
-        m_cancelled = true;
     }
 
-    bool is_cancelled() const
-    {
-        return m_cancelled;
-    }
+    Result& result;
+};
+
+// BackgroundAction has to inherit from Promise<Object> to allow it to be managed by EventLoop.
+// However, the internal result type is always ResultWrapper<Result>.
+template<typename Result>
+class BackgroundAction final : public Core::Promise<NonnullRefPtr<Core::Object>>
+    , private BackgroundActionBase {
+    C_OBJECT(BackgroundAction);
+
+public:
+    using Promise = Core::Promise<NonnullRefPtr<Core::Object>>;
 
     virtual ~BackgroundAction() = default;
 
+    Optional<Result> const& result() const { return m_result; }
+    Optional<Result>& result() { return m_result; }
+
 private:
-    BackgroundAction(Function<Result(BackgroundAction&)> action, Function<ErrorOr<void>(Result)> on_complete, Optional<Function<void(Error)>> on_error = {})
-        : Core::Object(&background_thread())
+    BackgroundAction(Function<Result(BackgroundAction&)> action, Function<ErrorOr<void>(Result&)> on_resolved, Optional<Function<void(Error)>> on_error = {})
+        : Promise(&background_thread())
         , m_action(move(action))
-        , m_on_complete(move(on_complete))
     {
+        if (on_resolved) {
+            this->on_resolved = [on_resolved = move(on_resolved)](auto& wrapped_result) {
+                return on_resolved(static_ptr_cast<ResultWrapper<Result>>(wrapped_result)->result);
+            };
+        }
+
         if (on_error.has_value())
             m_on_error = on_error.release_value();
 
+        if (this->on_resolved)
+            Core::EventLoop::current().add_job(*this);
+
         enqueue_work([this, origin_event_loop = &Core::EventLoop::current()] {
-            m_result = m_action(*this);
-            if (m_on_complete) {
-                origin_event_loop->deferred_invoke([this] {
-                    auto maybe_error = m_on_complete(m_result.release_value());
-                    if (maybe_error.is_error())
-                        m_on_error(maybe_error.release_error());
-                    remove_from_parent();
-                });
-                origin_event_loop->wake();
+            this->m_result = m_action(*this);
+            this->m_pending = ResultWrapper<Result>::construct(m_result.value());
```

Comment:
Why does this need to use the private API of Promise? I'm not seeing the requirement here. It should be enough to own a Promise and do any translations or call Promise::map or whatever after awaiting the promise here.

Reply:
I don't understand your suggestion

Reply:
Something like this (untested)
```c++
template<typename Result>
class BackgroundAction : public Core::Object
    , private BackgroundActionBase {
    C_OBJECT(BackgroundAction);

public:
    using Promise = Core::Promise<NonnullRefPtr<Core::Object>>;

    virtual ~BackgroundAction() = default;

    Optional<Result> const& result() const { return m_result; }
    Optional<Result>& result() { return m_result; }

    void cancel() { m_cancelled = true; }

private:
    BackgroundAction(Function<Result(BackgroundAction&)> action, Function<ErrorOr<void>(Result&)> on_resolved, Optional<Function<void(Error)>> on_error = {})
        : Core::Object(&background_thread())
        , m_action(move(action))
        , m_on_complete(move(on_resolved))
    {
        if (on_error.has_value())
            m_on_error = on_error.release_value();

        if (m_on_complete) {
            m_promise = Promise::try_create().release_value_but_fixme_should_propagate_errors();
            m_promise->on_resolved = [](auto& object) -> ErrorOr<void> {
                auto& self = verify_cast<BackgroundAction>(object);
                if (auto& err = self.m_on_complete(self.m_result.value()); err.is_error()) {
                    self.m_on_error(err.release_error());
                }
                return {};
            };
            Core::EventLoop::current().add_job(m_promise);
        }

        enqueue_work([this, origin_event_loop = &Core::EventLoop::current()] {
            this->m_result = m_action(*this);
            if (this->m_promise) {
                if (!this->m_cancelled) {
                    origin_event_loop->deferred_invoke([this] {
                        this->m_promise->resolve(*this);
                        // clean up m_promise somehow
                        this->remove_from_parent();
                    });
                    origin_event_loop->wake();
                }
                else {
                    this->m_promise->cancel();
                    if (this->m_on_error) {
                        this->m_on_error(Error::from_string_literal("Shouldn't have posted your hot takes on Twitter 5 years ago"));
                    }
                    this->remove_from_parent();
                }
            }
            else {
                 if (this->m_on_error) {
                        this->m_on_error(Error::from_string_literal("Shouldn't have posted your hot takes on Twitter 5 years ago"));
                }
                this->remove_from_parent();
            }
        });
    }

    Function<Result(BackgroundAction&)> m_action;
    Function<ErrorOr<void>(Result)> m_on_complete;
    Function<void(Error)> m_on_error = [](Error error) {
        dbgln("Error occurred while running a BackgroundAction: {}", error);
    };
    NonnullRefPtr<Promise> m_promise;
    Optional<Result> m_result;
    bool m_cancelled = false;
};
```